### PR TITLE
New form logic - Business Details

### PR DESCRIPTION
**Pull request by MorayM**
## Overview

Update the Business Details section to use a new form which removes the need to number questions. The order is now solely derived from `FieldConfiguration.cs` so there is no need to number views and take `id` parameters in `POST` actions. In addition, most names of fields and actions can be got from `nameof`, making the form a lot easier to refactor.

The main downside is that the generic `GET` actions must be replaced with specific actions for each page, but like the existing `POST` actions, their contents can for the most part be genericised - see `OrganisationDetailsController.cs` for more.

## Known issues

 - As the action used by the back buttons is generic across the entire form, it won't work for new sections until the whole form has been migrated.
 - There is no coherent form for URL routes yet - again this is something that is easier to work out once the whole form has been migrated.
 - All parts of the form that will eventually be removed have been tagged as `Obsolete`. They will of course still work, but beware if you have Warnings turned on in VS!

### CODE REVIEW

File: GLAA.Web/Controllers/DefaultController.cs [hunk 1]
```diff
@@ -64,17 +72,57 @@ protected virtual string GetLastViewPath(FormSection section)
             return View(viewPath, viewModel);
         }
 
+        /// <summary>
+        /// Return a view for the requested action name. If this view can't be viewed, it will seek backwards 
+        /// through the pages until one that can be viewed is found. If no view can be found then the last view
```
Comment (dougajmcdonald) [inline GLAA.Web/Controllers/DefaultController.cs:77]: Last view as in the one which was just viewed? or the last in the section?

File: GLAA.Web/Controllers/DefaultController.cs [hunk 2]
```diff
@@ -83,27 +131,103 @@ protected virtual string GetLastViewPath(FormSection section)
...
+        /// <summary>
+        /// Returns a <see cref="RedirectToActionResult"/> for the specified action name with the 
+        /// <code>back</code> direction indicator in the query string.
+        /// </summary>
+        /// <param name="section">The form section.</param>
+        /// <param name="actionName">The action name.</param>
+        /// <returns>A <see cref="RedirectToActionResult"/> for the specified action name with the <code>back</code> indicator.</returns>
+        protected virtual IActionResult RedirectBackToAction(FormSection section, string actionName)
+        {
+            return RedirectToAction(actionName, section.ToString(), new {back = true});
```
Comment (dougajmcdonald) [inline GLAA.Web/Controllers/DefaultController.cs:214]: Half of me wonders whether we can just use javascript for the back button and do-away with a lot of this!

File: GLAA.Web/Controllers/OrganisationDetailsController.cs [hunk 3]
```diff
@@ -95,75 +108,131 @@ public IActionResult AddPreviousTradingName(BusinessNameViewModel model)
...
         {
-            return OrganisationDetailsPost(model, 8);
+            return OrganisationDetailsPost(model, nameof(CommunicationPreference));
+        }
+
+        [HttpGet]
+        [ImportModelState]
+        public IActionResult LegalStatus(bool? back = false)
+        {
+            return OrganisationDetailsGet(nameof(LegalStatus), back);
```
Comment (dougajmcdonald) [inline GLAA.Web/Controllers/OrganisationDetailsController.cs:202]: The more I see these get actions the more I think that's totally fine. I don't we can keep things totally generic and we're finding more and more edge cases and complexity with trying to centralise it all. 
As a result, fully support breaking these down to separate actions.

File: GLAA.Web/Controllers/OrganisationDetailsController.cs [hunk 4]
```diff
@@ -182,32 +251,53 @@ public IActionResult AddPAYENumber(BusinessCredentialsViewModel model)
...
+        [ImportModelState]
+        public IActionResult VATStatus(bool? back = false)
+        {
+            return OrganisationDetailsGet(nameof(VATStatus), back);
         }
 
         [HttpPost]
         [ExportModelState]
-        public IActionResult SaveVATStatus(VATStatusViewModel model)
+        public IActionResult VATStatus(VATStatusViewModel model)
```
Comment (dougajmcdonald) [inline GLAA.Web/Controllers/OrganisationDetailsController.cs:266]: I like the removal of the `Save` prefix from these actions, feels more RESTy

File: GLAA.Web/FormLogic/IFieldConfiguration.cs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 using System.Collections.Generic;
-using GLAA.Web.Controllers;
 
 namespace GLAA.Web.FormLogic
 {
     public interface IFieldConfiguration
     {
         IDictionary<FormSection, FormPageDefinition[]> Fields { get; set; }
+
+        FormPageDefinition[] this[FormSection section] { get; }
```
Comment (dougajmcdonald) [inline GLAA.Web/FormLogic/IFieldConfiguration.cs:9]: I've never seen this syntax before, what's it doing?

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs [hunk 6]
```diff
@@ -38,11 +52,115 @@ public bool CanViewNextModel<TParent>(FormSection section, int id, TParent paren
...
-            return fieldConfiguration.Fields[section].Length;
+            return fieldConfiguration[section].Length;
         }
 
+        public FormPageDefinition GetLastPage(FormSection section)
+        {
+            return fieldConfiguration[section].Last();
+        }
+
+        public bool IsNextPageLastPage(FormSection section, string actionName)
```
Comment (dougajmcdonald) [inline GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:79]: huh? NextPageLastPage? Sounds a little confusing!

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs [hunk 7]
```diff
@@ -38,11 +52,115 @@ public bool CanViewNextModel<TParent>(FormSection section, int id, TParent paren
...
+        {
+            return fieldConfiguration[section][GetSectionLength(section) - 2].MatchesName(actionName);
```
Comment (dougajmcdonald) [inline GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:81]: What's `-2` here? could we name it as a variable. Don't mean to be picky but seen so much code over the years using magic numbers which becomes hard to reason about later on (like people just going, oh I need it one more than this in my specific situation, I'll make it -1) and then the world ends!

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs [hunk 8]
```diff
@@ -38,11 +52,115 @@ public bool CanViewNextModel<TParent>(FormSection section, int id, TParent paren
...
+
+            // if in doubt, return the final (summary) screen
+            return GetLastPage(section);
+        }
+
+        public FormPageDefinition GetNextPossiblePage<TParent>(FormSection section, string actionName, TParent parent)
+        {
+            var matchLocation = 0;
+            // Search through the list of pages for this section until we find the current one
+            for (var i = 0; i < GetSectionLength(section) - 1; i++)
```
Comment (dougajmcdonald) [inline GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:120]: Could this be <= and avoid the -1?

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs [hunk 9]
```diff
@@ -38,11 +52,115 @@ public bool CanViewNextModel<TParent>(FormSection section, int id, TParent paren
...
+            }
+
+            // if in doubt, return the final (summary) screen
+            return GetLastPage(section);
+        }
+
+        public FormPageDefinition GetPreviousPage(FormSection section, string actionName)
+        {
+            // Counting back from the end, find the page that matches this action name
+            for (var i = GetSectionLength(section) - 1; i > 0; i--)
```
Comment (dougajmcdonald) [inline GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:103]: Could this be >= and avoid the -1?

File: GLAA.Web/Views/Licence/TaskList.cshtml [hunk 10]
```diff
@@ -40,14 +40,14 @@
 
                                 @if (Model.OrganisationDetails.IsValid)
                                 {
-                                    <a class="task-name" href="@Url.Action("Part", "OrganisationDetails", new { id = 14 })" aria-describedby="company-information-completed">
+                                    <a class="task-name" href="@Url.Action("Summary", "OrganisationDetails")" aria-describedby="company-information-completed">
```
Comment (dougajmcdonald) [inline GLAA.Web/Views/Licence/TaskList.cshtml:43]: This makes me weep with joy! no more stupid numbering things!

File: GLAA.Web/Views/OrganisationDetails/CommunicationPreference.cshtml [hunk 11]
```diff
@@ -4,9 +4,9 @@
 
 @model GLAA.ViewModels.LicenceApplication.CommunicationPreferenceViewModel
```
Comment (dougajmcdonald) [inline GLAA.Web/Views/OrganisationDetails/CommunicationPreference.cshtml:5]: These file renames are going to save a lot of time trying to figure out which fields are on which pages!
+10

**Review by dougajmcdonald — APPROVED** [review]
This is great work, really like all the changes. Couple of nitpicks but nothing which should stop it going in. +1

File: GLAA.Web/Controllers/DefaultController.cs (hunk 1 above)
Comment (MorayM) [reply GLAA.Web/Controllers/DefaultController.cs:77]: I think I've been universal in my use of "last" as being "last in the section". The intention is that this usage mirrors Linq's `.Last()`.

File: GLAA.Web/Controllers/DefaultController.cs (hunk 2 above)
Comment (MorayM) [reply GLAA.Web/Controllers/DefaultController.cs:214]: That would make me SO HAPPY. The back logic is a massive pain, and the partial view that renders it makes my eyes hurt!

File: GLAA.Web/Controllers/OrganisationDetailsController.cs (hunk 3 above)
Comment (MorayM) [reply GLAA.Web/Controllers/OrganisationDetailsController.cs:202]: Agreed, it'll give us a lot more flexibility going forward.

File: GLAA.Web/FormLogic/IFieldConfiguration.cs (hunk 5 above)
Comment (MorayM) [reply GLAA.Web/FormLogic/IFieldConfiguration.cs:9]: I added this as shorthand to manipulate `Fields` a little easier. For future reference, this adds an index operator to the class - see here for more: https://stackoverflow.com/questions/424669/how-do-i-overload-the-operator-in-c-sharp

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs (hunk 6 above)
Comment (MorayM) [reply GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:79]: Yeah, I'm going to go through and rename everything once I'm done changing things. There's a few bits like this that make my brain hurt!

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs (hunk 7 above)
Comment (MorayM) [reply GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:81]: Fair point. The item at index `length - 2` is the penultimate item in the list, where the final item is usually the summary screen. I'll add some comments to this method to explain what's going on as it's not immediately obvious how it works (and the name doesn't help!)

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs (hunk 8 above)
Comment (MorayM) [reply GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:120]: Good spot, the `-1` shouldn't be there as the `<` will ensure we don't get index out of range.

File: GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs (hunk 9 above)
Comment (MorayM) [reply GLAA.Web/FormLogic/LicenceApplicationFormDefinition.cs:103]: Not in this case, we're counting backwards from the last item in the list so we're going to have to use it somewhere ☹️

**Pull request merged**